app/client/components/RegionFocusSwitcher.ts

```diff
@@ -0,0 +1,684 @@
+    getPanelElement('top') ? {type: 'panel', id: 'top'} as PanelRegion : null,
+  ].filter((x): x is PanelRegion => Boolean(x));
+
+  // If there is no doc with layout, just cycle through panels
+  if (!gristDoc) {
+    return [
+      ...commonPanels,
+      getPanelElement('main') ? {type: 'panel', id: 'main'} as PanelRegion : null,
+    ].filter((x): x is PanelRegion => Boolean(x));
+  }
+
+  // If there is a doc, also cycle through section ids
+  return [
+    ...gristDoc.viewLayout?.layout.getAllLeafIds().map(id => ({type: 'section', id} as SectionRegion)) ?? [],
+    ...commonPanels,
+  ];
+};
+
+/**
+ * Get the sibling region to focus in the regions given, compared to the current region and the direction.
+ *
+ * Exceptions:
+ *   - If we happen to be on the creator panel, focus back to the view layout active section,
+ *   - If we don't find anything, focus the first region in the cycle.
+ */
+const getSibling = (
+  current: Region | undefined,
+  regions: Region[],
+  direction: 'next' | 'prev',
+  gristDoc: GristDoc | null
+): Region | undefined => {
+  const isCreatorPanel = current?.type === 'panel' && current.id === 'right';
+
+  // First normally try to get current region in the cycle
+  let currentIndexInCycle = findRegionIndex(regions, current);
+
+  // If it's not found, it certainly means there is no current region set yet.
+  // In case of a grist doc, we can use the active section id as the "current index"
+  if ((currentIndexInCycle === -1 || isCreatorPanel) && gristDoc) {
+    currentIndexInCycle = findRegionIndex(regions, {type: 'section', id: gristDoc.viewModel.activeSectionId()});
+  }
+  // If we still don't find anything, it means we never set the current region before on a non-doc page,
+  // or we didn't find any current doc section. Return the first region as default.
+  if (currentIndexInCycle === -1) {
+    return regions[0];
+  }
+
+  // Normal case: just return the next or previous region in the cycle, wrapping around
+  const sibling = regions[mod(currentIndexInCycle + (direction === 'next' ? 1 : -1), regions.length)];
+  return sibling;
+};
+
+/**
+ * Blur the currently focused element in the given panel, if any.
+ */
+const blurPanelChild = (panel: PanelRegion) => {
+  const panelElement = getPanelElement(panel.id);
+  if (containsActiveElement(panelElement)) {
+    (document.activeElement as HTMLElement).blur();
+  }
+};
+
+let _focusLocked: {el: HTMLElement | null, cb: ((event: KeyboardEvent) => void) | null} = {el: null, cb: null};
+
+const disableFocusLock = () => {
+  const {el, cb} = _focusLocked;
+  if (el && cb) {
+    el.removeEventListener('keydown', cb);
+    _focusLocked = {el: null, cb: null};
+  }
+};
+
+const enableFocusLock = (panelElement: HTMLElement) => {
+  disableFocusLock();
+  const focusTrap = (event: KeyboardEvent) => {
+    if (event.key === 'Tab') {
+      trapTabKey(panelElement, event);
+    }
+  };
+  panelElement.addEventListener('keydown', focusTrap);
```

Comment:
Could you add comments what `enableFocusLock` does?

Optionally, I think you can simplify a little using grainjs, since it has tools just for this sort of thing (creating and later disposing some resource like an event listener):
```
const _focusLockHolder = Holder.create(null);

const disableFocusLock = () => _focusLockHolder.clear();

const enableFocusLock = (panelElement) => {
  _focusLockHolder.autoDispose(dom.onElem(panelElement, 'keydown', (event, elem) => {
    if (event.key === 'Tab') {
      trapTabKey(elem, event);
    }
  }));
};
```